src/Bicep.Core/TypeSystem/DeployTimeConstantIndirectViolationVisitor.cs

```diff
@@ -131,49 +148,58 @@ private void FlagIfAccessingEntireResourceOrModule(SyntaxBase syntax)
                 case not PropertyAccessSyntax and not ArrayAccessSyntax when
                     this.ResourceTypeResolver.TryResolveResourceOrModuleSymbolAndBodyType(syntax) is ({ } accessedSymbol, { } accessedBodyType):
                     {
-                        var variableDependencyChain = this.BuildVariablDependencyChain(accessedSymbol.Name);
-                        this.FlagDeployTimeConstantViolation(accessedSymbol, accessedBodyType, variableDependencyChain);
-
+                        this.FlagDeployTimeConstantViolationWithVariableDependencies(accessedSymbol, accessedBodyType);
                         break;
                     }
                 // var foo = [for x in [...]: {
                 //   bar: myVNets[1] <-- accessing an entire resource/module via an array index.
                 // }]
-                case ArrayAccessSyntax { IndexExpression: IntegerLiteralSyntax } arrayAccessSyntax when
+                case ArrayAccessSyntax arrayAccessSyntax when
+                    arrayAccessSyntax.BaseExpression == syntax &&
                     this.SemanticModel.Binder.GetParent(arrayAccessSyntax) is not PropertyAccessSyntax and not ArrayAccessSyntax &&
-                    this.ResourceTypeResolver.TryResolveResourceOrModuleSymbolAndBodyType(syntax) is ({ } accessedSymbol, { } accessedBodyType):
+                    this.ResourceTypeResolver.TryResolveResourceOrModuleSymbolAndBodyType(arrayAccessSyntax) is ({ } resourceSymbol, { } resourceType):
+                {
+                    var arrayIndexExprType = this.SemanticModel.GetTypeInfo(arrayAccessSyntax.IndexExpression);
+                    if (arrayIndexExprType.IsIntegerOrIntegerLiteral())
                     {
-                        var variableDependencyChain = this.BuildVariablDependencyChain(accessedSymbol.Name);
-                        this.FlagDeployTimeConstantViolation(accessedSymbol, accessedBodyType, variableDependencyChain);
-
-                        break;
+                        this.FlagDeployTimeConstantViolationWithVariableDependencies(resourceSymbol, resourceType);
                     }
```

Comment:
I'm not sure I follow the logic here. Are other index expression types *not* considered deploy time constant violations? (E.g., an index of type `any` or whose type is a union of integer literals)

Reply:
The scope of the containing function is only supposed to flag entire resource accesses only, so in the case of `ArrayAccessSyntax` which is the syntax type for both `foo['property']` and `foos[0]`, it seems like it only makes sense to check for integer expressions on resource collections. The `VisitArrayAccessSyntax` method handles the string case of index expression. I think in other type cases it would be another type a diagnostic like "cannot use bool here". There's also a bit of a relationship between these conditions and how the `TryResourceOrModuleSymbolAndBodyType` resolves the resource symbol. It has more checking inside to align string/vs integer access against singular/collection resources. If they don't align, it resolves to null which fails these conditions. To help with my own confusion I wanted to be exhaustive in the test cases.

I will double check this.

Reply:
Fixed the gap of checking for `any` and `unions`.